### PR TITLE
msg: refactor how terminal messages are printed

### PR DESCRIPTION
**Pull request by kasper93**

### CODE REVIEW

**Comment by kasper93** [comment]
marking as draft, because need to change how lines are prefixed still

**Review by guidocella — APPROVED** [review]

File: common/msg.c [hunk 1]
```diff
@@ -299,42 +306,60 @@ static bool test_terminal_level(struct mp_log *log, int lev)
            !(lev == MSGL_STATUS && terminal_in_background());
 }
 
-static void print_terminal_line(struct mp_log *log, int lev,
-                                char *text,  char *trail)
-{
-    if (!test_terminal_level(log, lev))
-        return;
+// This is very basic way to infer needed width for a string.
+int term_disp_width(const char *str) {
```
Comment (Dudemanguy) [inline common/msg.c:310]: Should be static here.

**Review by N-R-K — COMMENTED** [review]
The "refactor how terminal messages are printed" commit is a bit too big to review thoroughly. But at a glance, nothing obviously wrong stands out (aside from the function which should be `static`).

Also, all those `*_append` functions are quadratic [Schlemiels](https://en.wikipedia.org/wiki/Joel_Spolsky#Schlemiel_the_Painter.27s_algorithm). If there's already non-quadratic bstr functions available for this, I'd probably use those. If not, then leave it be, the string likely won't get big enough for it to really matter.

**Comment by christoph-heinrich** [comment]
> The "refactor how terminal messages are printed" commit is a bit too big to review thoroughly. But at a glance, nothing obviously wrong stands out (aside from the function which should be `static`).
> 
> Also, all those `*_append` functions are quadratic [Schlemiels](https://en.wikipedia.org/wiki/Joel_Spolsky#Schlemiel_the_Painter.27s_algorithm). If there's already non-quadratic bstr functions available for this, I'd probably use those. If not, then leave it be, the string likely won't get big enough for it to really matter.

There are `bstr_xappend` and `bstr_xappend_asprintf`, from a quick look I think that's all that's needed?

**Comment by kasper93** [comment]
> There are bstr_xappend() and bstr_xappend_asprintf(), from a quick look I think that's all that's needed?

Yes, don't worry, I will change that.

**Comment by llyyr** [comment]
The current iteration of this PR fixes the functional issues with autocomplete. I don't see anything wrong now as far as functionality goes.

**Comment by Dudemanguy** [comment]
Overall looks OK. The only thing I personally don't like is the status line always being cleared on exit instead of being kept.

**Comment by kasper93** [comment]
rebased

> Overall looks OK. The only thing I personally don't like is the status line always being cleared on exit instead of being kept.

**Comment by kasper93** [comment]
Yes, it will does now. It is "virtual" status line that only exists when there is actual status line to display and cleared otherwise. Keeping it on exits, doesn't fit current design.

**Comment by kasper93** [comment]
Rebased after mp_thread merge.

**Review by Dudemanguy — APPROVED** [review]
Not really qualified on terminal stuff but looks fine.

**Pull request merged**